Should rawConfig include policy name?

#5397 assumed that rawConfig includes policy name, but the constructor 
`LbConfig(String policyName, Map<String, ?> rawConfigValue)` is never passing `rawConfigValue` with the policy name. That ends up a bug

```
FINEST: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Resolved address: [[[/0:0:0:0:0:0:0:1:21607]/{io.grpc.grpclb.lbAddrAuthority=fake-authority-that-is-always-the-same}]], config={service-config={loadBalancingConfig=[{xds_experimental={balancerName=[::1]:21607}}, {grpclb={}}, {xds={balancerName=[::1]:21607}}], healthCheckConfig={serviceName=}}}
May 29, 2019 5:11:20 PM io.grpc.ChannelLogger log
FINEST: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Address resolved: [[[/0:0:0:0:0:0:0:1:21607]/{io.grpc.grpclb.lbAddrAuthority=fake-authority-that-is-always-the-same}]]
May 29, 2019 5:11:20 PM io.grpc.internal.ClientCallImpl logIfContextNarrowedTimeout
FINE: Call timeout set to '4998908748' ns, due to context deadline. Explicit call timeout was not set.
May 29, 2019 5:11:20 PM io.grpc.ChannelLogger log
FINEST: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Service config changed
May 29, 2019 5:11:20 PM io.grpc.internal.ClientCallImpl logIfContextNarrowedTimeout
FINE: Call timeout set to '4997013555' ns, due to context deadline. Explicit call timeout was not set.
May 29, 2019 5:11:20 PM io.grpc.ChannelLogger log
FINEST: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Load balancer changed from PickFirstLoadBalancer to XdsLoadBalancer
May 29, 2019 5:11:20 PM io.grpc.ChannelLogger log
FINEST: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Load-balancing config: {balancerName=[::1]:21607}
May 29, 2019 5:11:20 PM io.grpc.internal.ManagedChannelImpl$1 uncaughtException
SEVERE: [Channel<7>: (test_service.test_namespace.gslb.googleprod.com)] Uncaught exception in the SynchronizationContext. Panic!
io.grpc.StatusRuntimeException: UNKNOWN: Failed to parse config value '[::1]:21607' for key 'balancerName' in '{balancerName=[::1]:21607}' is not object
	at io.grpc.Status.asRuntimeException(Status.java:524)
	at io.grpc.xds.XdsLoadBalancer.handleResolvedAddresses(XdsLoadBalancer.java:133)
	at io.grpc.internal.AutoConfiguredLoadBalancerFactory$AutoConfiguredLoadBalancer.handleResolvedAddresses(AutoConfiguredLoadBalancerFactory.java:152)
	at io.grpc.internal.ManagedChannelImpl$NameResolverListener$1NamesResolved.run(ManagedChannelImpl.java:1346)
	at io.grpc.SynchronizationContext.drain(SynchronizationContext.java:88)
	at io.grpc.SynchronizationContext.execute(SynchronizationContext.java:120)
	at com.google.net.grpc.nameresolver.EnvelopeNameResolver$EnvelopeStreamListener.onResponse(EnvelopeNameResolver.java:306)
	at com.google.net.rpc.messenger.NettyMessengerClient$Stream$1Receiver.run(NettyMessengerClient.java:265)
	at com.google.common.util.concurrent.SequentialExecutor$1.run(SequentialExecutor.java:120)
	at com.google.common.util.concurrent.SequentialExecutor$QueueWorker.workOnQueue(SequentialExecutor.java:227)
	at com.google.common.util.concurrent.SequentialExecutor$QueueWorker.run(SequentialExecutor.java:171)
	at com.google.net.eventmanager.AbstractFutureTask$Sync.innerRun(AbstractFutureTask.java:260)
	at com.google.net.eventmanager.AbstractFutureTask.run(AbstractFutureTask.java:121)
	at com.google.net.eventmanager.EventManagerImpl.runTask(EventManagerImpl.java:618)
	at com.google.net.eventmanager.EventManagerImpl.internalRunWorkerLoop(EventManagerImpl.java:1036)
	at com.google.net.eventmanager.EventManagerImpl.runWorkerLoop(EventManagerImpl.java:916)
	at com.google.net.eventmanager.WorkerThreadInfo.runWorkerLoop(WorkerThreadInfo.java:160)
	at com.google.net.eventmanager.EventManagerImpl$WorkerThread.run(EventManagerImpl.java:1903)
Caused by: java.lang.ClassCastException: value '[::1]:21607' for key 'balancerName' in '{balancerName=[::1]:21607}' is not object
	at io.grpc.internal.ServiceConfigUtil.getObject(ServiceConfigUtil.java:458)
	at io.grpc.internal.ServiceConfigUtil.unwrapLoadBalancingConfig(ServiceConfigUtil.java:364)
	at io.grpc.xds.XdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(XdsLoadBalancerProvider.java:77)
	at io.grpc.xds.XdsLoadBalancer.handleResolvedAddresses(XdsLoadBalancer.java:131)
	... 16 more
```

--- a/core/src/main/java/io/grpc/internal/ServiceConfigUtil.java
+++ b/core/src/main/java/io/grpc/internal/ServiceConfigUtil.java
@@ -379,18 +379,16 @@ public final class ServiceConfigUtil {
   /**
    * Extracts the loadbalancer name from xds loadbalancer config.
    */
-  public static String getBalancerNameFromXdsConfig(LbConfig xdsConfig) {
-    Map<String, ?> map = xdsConfig.getRawConfigValue();
-    return getString(map, XDS_CONFIG_BALANCER_NAME_KEY);
+  public static String getBalancerNameFromXdsConfig(Map<String, ?> rawXdsConfig) {
+    return getString(rawXdsConfig, XDS_CONFIG_BALANCER_NAME_KEY);
   }
 
   /**
    * Extracts list of child policies from xds loadbalancer config.
    */
   @Nullable
-  public static List<LbConfig> getChildPolicyFromXdsConfig(LbConfig xdsConfig) {
-    Map<String, ?> map = xdsConfig.getRawConfigValue();
-    List<?> rawChildPolicies = getList(map, XDS_CONFIG_CHILD_POLICY_KEY);
+  public static List<LbConfig> getChildPolicyFromXdsConfig(Map<String, ?> rawXdsConfig) {
+    List<?> rawChildPolicies = getList(rawXdsConfig, XDS_CONFIG_CHILD_POLICY_KEY);
     if (rawChildPolicies != null) {
       return unwrapLoadBalancingConfigList(checkObjectList(rawChildPolicies));
     }
@@ -401,9 +399,8 @@ public final class ServiceConfigUtil {
    * Extracts list of fallback policies from xds loadbalancer config.
    */
   @Nullable
-  public static List<LbConfig> getFallbackPolicyFromXdsConfig(LbConfig xdsConfig) {
-    Map<String, ?> map = xdsConfig.getRawConfigValue();
-    List<?> rawFallbackPolicies = getList(map, XDS_CONFIG_FALLBACK_POLICY_KEY);
+  public static List<LbConfig> getFallbackPolicyFromXdsConfig(Map<String, ?> rawXdsConfig) {
+    List<?> rawFallbackPolicies = getList(rawXdsConfig, XDS_CONFIG_FALLBACK_POLICY_KEY);
     if (rawFallbackPolicies != null) {
       return unwrapLoadBalancingConfigList(checkObjectList(rawFallbackPolicies));
     }

--- a/core/src/test/java/io/grpc/internal/ServiceConfigUtilTest.java
+++ b/core/src/test/java/io/grpc/internal/ServiceConfigUtilTest.java
@@ -34,75 +34,74 @@ import org.junit.runners.JUnit4;
 public class ServiceConfigUtilTest {
   @Test
   public void getBalancerNameFromXdsConfig() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"round_robin\" : {}}, {\"lbPolicy2\" : {\"key\" : \"val\"}}],"
         + "\"fallbackPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
     assertEquals(
         "dns:///balancer.example.com:8080",
-        ServiceConfigUtil.getBalancerNameFromXdsConfig(
-            ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig)))));
+        ServiceConfigUtil.getBalancerNameFromXdsConfig(checkObject(JsonParser.parse(rawLbConfig))));
   }
 
   @Test
   public void getChildPolicyFromXdsConfig() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"round_robin\" : {}}, {\"lbPolicy2\" : {\"key\" : \"val\"}}],"
         + "\"fallbackPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
     LbConfig expectedChildPolicy1 = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"round_robin\" : {}}")));
     LbConfig expectedChildPolicy2 = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"lbPolicy2\" : {\"key\" : \"val\"}}")));
 
     List<LbConfig> childPolicies = ServiceConfigUtil.getChildPolicyFromXdsConfig(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))));
+        checkObject(JsonParser.parse(rawLbConfig)));
 
     assertThat(childPolicies).containsExactly(expectedChildPolicy1, expectedChildPolicy2);
   }
 
   @Test
   public void getChildPolicyFromXdsConfig_null() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"fallbackPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
 
     List<LbConfig> childPolicies = ServiceConfigUtil.getChildPolicyFromXdsConfig(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))));
+        checkObject(JsonParser.parse(rawLbConfig)));
 
     assertThat(childPolicies).isNull();
   }
 
   @Test
   public void getFallbackPolicyFromXdsConfig() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"round_robin\" : {}}, {\"lbPolicy2\" : {\"key\" : \"val\"}}],"
         + "\"fallbackPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
     LbConfig expectedFallbackPolicy1 = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"lbPolicy3\" : {\"key\" : \"val\"}}")));
     LbConfig expectedFallbackPolicy2 = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"lbPolicy4\" : {}}")));
 
     List<LbConfig> childPolicies = ServiceConfigUtil.getFallbackPolicyFromXdsConfig(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))));
+        checkObject(JsonParser.parse(rawLbConfig)));
 
     assertThat(childPolicies).containsExactly(expectedFallbackPolicy1, expectedFallbackPolicy2);
   }
 
   @Test
   public void getFallbackPolicyFromXdsConfig_null() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"round_robin\" : {}}, {\"lbPolicy2\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
 
     List<LbConfig> fallbackPolicies = ServiceConfigUtil.getFallbackPolicyFromXdsConfig(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))));
+        checkObject(JsonParser.parse(rawLbConfig)));
 
     assertThat(fallbackPolicies).isNull();
   }

--- a/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java
+++ b/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java
@@ -42,6 +42,7 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
 
   private static final LbConfig DEFAULT_FALLBACK_POLICY =
       new LbConfig("round_robin", ImmutableMap.<String, Void>of());
+  private static final String XDS_POLICY_NAME = "xds_experimental";
 
   @Override
   public boolean isAvailable() {
@@ -55,7 +56,7 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
 
   @Override
   public String getPolicyName() {
-    return "xds_experimental";
+    return XDS_POLICY_NAME;
   }
 
   @Override
@@ -73,8 +74,7 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
   static ConfigOrError parseLoadBalancingConfigPolicy(
       Map<String, ?> rawLoadBalancingPolicyConfig, LoadBalancerRegistry registry) {
     try {
-      LbConfig newLbConfig =
-          ServiceConfigUtil.unwrapLoadBalancingConfig(rawLoadBalancingPolicyConfig);
+      LbConfig newLbConfig = new LbConfig(XDS_POLICY_NAME, rawLoadBalancingPolicyConfig);
       String newBalancerName = ServiceConfigUtil.getBalancerNameFromXdsConfig(newLbConfig);
       LbConfig childPolicy = selectChildPolicy(newLbConfig, registry);
       LbConfig fallbackPolicy = selectFallbackPolicy(newLbConfig, registry);

--- a/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java
+++ b/xds/src/main/java/io/grpc/xds/XdsLoadBalancerProvider.java
@@ -42,7 +42,6 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
 
   private static final LbConfig DEFAULT_FALLBACK_POLICY =
       new LbConfig("round_robin", ImmutableMap.<String, Void>of());
-  private static final String XDS_POLICY_NAME = "xds_experimental";
 
   @Override
   public boolean isAvailable() {
@@ -56,7 +55,7 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
 
   @Override
   public String getPolicyName() {
-    return XDS_POLICY_NAME;
+    return "xds_experimental";
   }
 
   @Override
@@ -74,10 +73,10 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
   static ConfigOrError parseLoadBalancingConfigPolicy(
       Map<String, ?> rawLoadBalancingPolicyConfig, LoadBalancerRegistry registry) {
     try {
-      LbConfig newLbConfig = new LbConfig(XDS_POLICY_NAME, rawLoadBalancingPolicyConfig);
-      String newBalancerName = ServiceConfigUtil.getBalancerNameFromXdsConfig(newLbConfig);
-      LbConfig childPolicy = selectChildPolicy(newLbConfig, registry);
-      LbConfig fallbackPolicy = selectFallbackPolicy(newLbConfig, registry);
+      String newBalancerName =
+          ServiceConfigUtil.getBalancerNameFromXdsConfig(rawLoadBalancingPolicyConfig);
+      LbConfig childPolicy = selectChildPolicy(rawLoadBalancingPolicyConfig, registry);
+      LbConfig fallbackPolicy = selectFallbackPolicy(rawLoadBalancingPolicyConfig, registry);
       return ConfigOrError.fromConfig(new XdsConfig(newBalancerName, childPolicy, fallbackPolicy));
     } catch (RuntimeException e) {
       return ConfigOrError.fromError(
@@ -86,16 +85,20 @@ public final class XdsLoadBalancerProvider extends LoadBalancerProvider {
   }
 
   @VisibleForTesting
-  static LbConfig selectFallbackPolicy(LbConfig lbConfig, LoadBalancerRegistry lbRegistry) {
-    List<LbConfig> fallbackConfigs = ServiceConfigUtil.getFallbackPolicyFromXdsConfig(lbConfig);
+  static LbConfig selectFallbackPolicy(
+      Map<String, ?> rawLoadBalancingPolicyConfig, LoadBalancerRegistry lbRegistry) {
+    List<LbConfig> fallbackConfigs =
+        ServiceConfigUtil.getFallbackPolicyFromXdsConfig(rawLoadBalancingPolicyConfig);
     LbConfig fallbackPolicy = selectSupportedLbPolicy(fallbackConfigs, lbRegistry);
     return fallbackPolicy == null ? DEFAULT_FALLBACK_POLICY : fallbackPolicy;
   }
 
   @Nullable
   @VisibleForTesting
-  static LbConfig selectChildPolicy(LbConfig lbConfig, LoadBalancerRegistry lbRegistry) {
-    List<LbConfig> childConfigs = ServiceConfigUtil.getChildPolicyFromXdsConfig(lbConfig);
+  static LbConfig selectChildPolicy(
+      Map<String, ?> rawLoadBalancingPolicyConfig, LoadBalancerRegistry lbRegistry) {
+    List<LbConfig> childConfigs =
+        ServiceConfigUtil.getChildPolicyFromXdsConfig(rawLoadBalancingPolicyConfig);
     return selectSupportedLbPolicy(childConfigs, lbRegistry);
   }
 

--- a/xds/src/test/java/io/grpc/xds/XdsLoadBalancerProviderTest.java
+++ b/xds/src/test/java/io/grpc/xds/XdsLoadBalancerProviderTest.java
@@ -106,19 +106,18 @@ public class XdsLoadBalancerProviderTest {
 
   @Test
   public void selectChildPolicy() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported_1\" : {}}, {\"supported_1\" : {\"key\" : \"val\"}},"
         + "{\"supported_2\" : {\"key\" : \"val\"}}],"
         + "\"fallbackPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
     LbConfig expectedChildPolicy =
         ServiceConfigUtil.unwrapLoadBalancingConfig(
             checkObject(JsonParser.parse("{\"supported_1\" : {\"key\" : \"val\"}}")));
 
     LbConfig childPolicy =
-        XdsLoadBalancerProvider.selectChildPolicy(
-            ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))),
+        XdsLoadBalancerProvider.selectChildPolicy(checkObject(JsonParser.parse(rawLbConfig)),
             lbRegistry);
 
     assertEquals(expectedChildPolicy, childPolicy);
@@ -126,48 +125,45 @@ public class XdsLoadBalancerProviderTest {
 
   @Test
   public void selectFallBackPolicy() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"supported_1\" : {\"key\" : \"val\"}},"
         + "{\"supported_2\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     LbConfig expectedFallbackPolicy = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"supported_1\" : {\"key\" : \"val\"}}")));
 
     LbConfig fallbackPolicy = XdsLoadBalancerProvider.selectFallbackPolicy(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))),
-        lbRegistry);
+        checkObject(JsonParser.parse(rawLbConfig)), lbRegistry);
 
     assertEquals(expectedFallbackPolicy, fallbackPolicy);
   }
 
   @Test
   public void selectFallBackPolicy_roundRobinIsDefault() throws Exception {
-    String lbConfig = "{\"xds_experimental\" : { "
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
-        + "}}";
+        + "}";
     LbConfig expectedFallbackPolicy = ServiceConfigUtil.unwrapLoadBalancingConfig(
         checkObject(JsonParser.parse("{\"round_robin\" : {}}")));
 
     LbConfig fallbackPolicy = XdsLoadBalancerProvider.selectFallbackPolicy(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(
-            checkObject(JsonParser.parse(lbConfig))),
-        lbRegistry);
+        checkObject(JsonParser.parse(rawLbConfig)), lbRegistry);
 
     assertEquals(expectedFallbackPolicy, fallbackPolicy);
   }
 
   @Test
   public void parseLoadBalancingConfigPolicy() throws Exception {
-    String lbConfigRaw = "{"
+    String rawLbConfig = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"supported_1\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"round_robin\" : {\"key\" : \"val\"}},"
         + "{\"supported_2\" : {\"key\" : \"val\"}}]"
         + "}";
-    Map<String, ?> rawlbConfigMap = checkObject(JsonParser.parse(lbConfigRaw));
+    Map<String, ?> rawlbConfigMap = checkObject(JsonParser.parse(rawLbConfig));
     ConfigOrError configOrError =
         XdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(rawlbConfigMap, lbRegistry);
 

--- a/xds/src/test/java/io/grpc/xds/XdsLoadBalancerProviderTest.java
+++ b/xds/src/test/java/io/grpc/xds/XdsLoadBalancerProviderTest.java
@@ -16,15 +16,18 @@
 
 package io.grpc.xds;
 
+import static com.google.common.truth.Truth.assertThat;
 import static org.junit.Assert.assertEquals;
 
 import io.grpc.LoadBalancer;
 import io.grpc.LoadBalancer.Helper;
 import io.grpc.LoadBalancerProvider;
 import io.grpc.LoadBalancerRegistry;
+import io.grpc.NameResolver.ConfigOrError;
 import io.grpc.internal.JsonParser;
 import io.grpc.internal.ServiceConfigUtil;
 import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import io.grpc.xds.XdsLoadBalancer.XdsConfig;
 import java.util.Map;
 import org.junit.Before;
 import org.junit.Rule;
@@ -103,7 +106,7 @@ public class XdsLoadBalancerProviderTest {
 
   @Test
   public void selectChildPolicy() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfig = "{\"xds_experimental\" : { "
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported_1\" : {}}, {\"supported_1\" : {\"key\" : \"val\"}},"
         + "{\"supported_2\" : {\"key\" : \"val\"}}],"
@@ -115,7 +118,7 @@ public class XdsLoadBalancerProviderTest {
 
     LbConfig childPolicy =
         XdsLoadBalancerProvider.selectChildPolicy(
-            ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfigRaw))),
+            ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))),
             lbRegistry);
 
     assertEquals(expectedChildPolicy, childPolicy);
@@ -123,7 +126,7 @@ public class XdsLoadBalancerProviderTest {
 
   @Test
   public void selectFallBackPolicy() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfig = "{\"xds_experimental\" : { "
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"supported_1\" : {\"key\" : \"val\"}},"
@@ -133,7 +136,7 @@ public class XdsLoadBalancerProviderTest {
         checkObject(JsonParser.parse("{\"supported_1\" : {\"key\" : \"val\"}}")));
 
     LbConfig fallbackPolicy = XdsLoadBalancerProvider.selectFallbackPolicy(
-        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfigRaw))),
+        ServiceConfigUtil.unwrapLoadBalancingConfig(checkObject(JsonParser.parse(lbConfig))),
         lbRegistry);
 
     assertEquals(expectedFallbackPolicy, fallbackPolicy);
@@ -141,7 +144,7 @@ public class XdsLoadBalancerProviderTest {
 
   @Test
   public void selectFallBackPolicy_roundRobinIsDefault() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfig = "{\"xds_experimental\" : { "
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"lbPolicy4\" : {}}]"
         + "}}";
@@ -150,10 +153,34 @@ public class XdsLoadBalancerProviderTest {
 
     LbConfig fallbackPolicy = XdsLoadBalancerProvider.selectFallbackPolicy(
         ServiceConfigUtil.unwrapLoadBalancingConfig(
-            checkObject(JsonParser.parse(lbConfigRaw))),
+            checkObject(JsonParser.parse(lbConfig))),
         lbRegistry);
 
     assertEquals(expectedFallbackPolicy, fallbackPolicy);
+  }
+
+  @Test
+  public void parseLoadBalancingConfigPolicy() throws Exception {
+    String lbConfigRaw = "{"
+        + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
+        + "\"childPolicy\" : [{\"lbPolicy3\" : {\"key\" : \"val\"}}, {\"supported_1\" : {}}],"
+        + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"round_robin\" : {\"key\" : \"val\"}},"
+        + "{\"supported_2\" : {\"key\" : \"val\"}}]"
+        + "}";
+    Map<String, ?> rawlbConfigMap = checkObject(JsonParser.parse(lbConfigRaw));
+    ConfigOrError configOrError =
+        XdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(rawlbConfigMap, lbRegistry);
+
+    assertThat(configOrError.getError()).isNull();
+    assertThat(configOrError.getConfig()).isInstanceOf(XdsConfig.class);
+    assertThat(configOrError.getConfig()).isEqualTo(
+        new XdsConfig(
+            "dns:///balancer.example.com:8080",
+            ServiceConfigUtil.unwrapLoadBalancingConfig(
+                checkObject(JsonParser.parse("{\"supported_1\" : {}}"))),
+            ServiceConfigUtil.unwrapLoadBalancingConfig(
+                checkObject(JsonParser.parse("{\"round_robin\" : {\"key\" : \"val\"}}"))))
+    );
   }
 
   @SuppressWarnings("unchecked")

--- a/xds/src/test/java/io/grpc/xds/XdsLoadBalancerTest.java
+++ b/xds/src/test/java/io/grpc/xds/XdsLoadBalancerTest.java
@@ -295,11 +295,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void resolverEvent_standardModeToStandardMode() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -318,10 +318,10 @@ public class XdsLoadBalancerTest {
             ArgumentMatchers.<CallOptions>any());
 
 
-    lbConfigRaw = "{\"xds_experimental\" : { "
+    lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig2 = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig2).build();
@@ -346,11 +346,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void resolverEvent_standardModeToCustomMode() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -365,11 +365,11 @@ public class XdsLoadBalancerTest {
         .newCall(ArgumentMatchers.<MethodDescriptor<?, ?>>any(),
             ArgumentMatchers.<CallOptions>any());
 
-    lbConfigRaw = "{\"xds_experimental\" : { "
+    lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"supported_2\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig2 = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig2).build();
@@ -392,11 +392,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void resolverEvent_customModeToStandardMode() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"supported_2\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -413,11 +413,11 @@ public class XdsLoadBalancerTest {
 
     assertThat(lb.getXdsLbStateForTest().childPolicy).isNotNull();
 
-    lbConfigRaw = "{\"xds_experimental\" : { "
+    lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig2 = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig2).build();
@@ -440,11 +440,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void resolverEvent_customModeToCustomMode() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"supported_2\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -460,11 +460,11 @@ public class XdsLoadBalancerTest {
         .newCall(ArgumentMatchers.<MethodDescriptor<?, ?>>any(),
             ArgumentMatchers.<CallOptions>any());
 
-    lbConfigRaw = "{\"xds_experimental\" : { "
+    lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"fallback_1\" : {\"key\" : \"val\"}}, {\"unfallback_1\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig2 = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig2).build();
@@ -486,11 +486,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void resolverEvent_balancerNameChange() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -505,11 +505,11 @@ public class XdsLoadBalancerTest {
         .newCall(ArgumentMatchers.<MethodDescriptor<?, ?>>any(),
             ArgumentMatchers.<CallOptions>any());
 
-    lbConfigRaw = "{\"xds_experimental\" : { "
+    lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8443\","
         + "\"childPolicy\" : [{\"fallback_1\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig2 = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig2).build();
@@ -765,10 +765,10 @@ public class XdsLoadBalancerTest {
   }
 
   private static Attributes standardModeWithFallback1Attributes() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{"
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"fallbackPolicy\" : [{\"fallback_1\" : { \"fallback_1_option\" : \"yes\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     return Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();
@@ -776,11 +776,11 @@ public class XdsLoadBalancerTest {
 
   @Test
   public void shutdown_cleanupTimers() throws Exception {
-    String lbConfigRaw = "{\"xds_experimental\" : { "
+    String lbConfigRaw = "{ "
         + "\"balancerName\" : \"dns:///balancer.example.com:8080\","
         + "\"childPolicy\" : [{\"unsupported\" : {\"key\" : \"val\"}}, {\"unsupported_2\" : {}}],"
         + "\"fallbackPolicy\" : [{\"unsupported\" : {}}, {\"fallback_1\" : {\"key\" : \"val\"}}]"
-        + "}}";
+        + "}";
     @SuppressWarnings("unchecked")
     Map<String, ?> lbConfig = (Map<String, ?>) JsonParser.parse(lbConfigRaw);
     Attributes attrs = Attributes.newBuilder().set(ATTR_LOAD_BALANCING_CONFIG, lbConfig).build();

